### ISSUE TITLE
Remove Thread.interrupt() restriction.

### ISSUE THREAD

**Issue by inscriber** [zeromq/jeromq#116]
The wiki says:

"Context.term() must be called before interrupting a thread by Thread.interrupt. Otherwise internal pipe which is used to signal ZMQ commands will be shutdown unexpectedly."

However, in some cases this means that there is no clean way for an application to shut down when it is blocked on recv().

Consider a thread (thread B) that is blocked on a socket's recv().

Suppose the main thread (thread A) of the application wants to shut down, it needs to cleanly terminate thread B. This typically involves sending an interrupt to thread B to unblock, but wait! we need to call Context.term() first. However, the wiki says:

"Close all the sockets properly otherwise Context.term() will wait forever"

OK, we need to close the socket first, but wait! The ZMQ guide says:

"Isolate data privately within its thread and never share data in multiple threads. The only exception to this are ØMQ contexts, which are threadsafe."

Hmm, so the only safe way to close the socket is on the thread that is currently blocked (thread B), too bad we can't do that.

So now the only way for thread B to unblock is to wait for a new message or a timeout (which may have been set to no timeout). If there are no more messages and there is no timeout, the thread will be blocked indefinitely. But regardless, there should be a means to preemptively unblock the thread.

Even if we were to ignore the shutdown problem, this restriction also leads to more awkward code because now the context needs to be exposed to other threads where it might not have been necessary without the restriction.

In fact, if you look at the Java example in the ZMQ guide and even the examples in this repository, it breaks the interrupt rule.

This is a pretty common pattern you see throughout the code:

```
while (!Thread.currentThread ().isInterrupted ()) {
    ...
}
receiver.close ();
context.term ();
```

Notice how it expects an interrupt before closing the socket and context.

**Comment by virtuald** [comment]
+1 It would be great if jeromq was Thread.interrupt safe.

**Comment by zhivko** [comment]
+1 It would be great if jeromq was Thread.interrupt safe. :)

**Comment by barbaragomes** [comment]
+1 It would be great if jeromq was Thread.interrupt safe. :)

**Comment by zhivko** [comment]
+1

**Comment by daveyarwood** [comment]
> I understand this restriction of termination is awkward and needs to be revised fundamentally.

@miniway It's been a while -- wondering if you've had any further insights on this? I would be interested in taking a stab at a PR to implement a fix for this issue, but I could use some direction on what the best approach might be.

**Comment by miniway** [comment]
I'm sorry that I have to say I haven't had time to look into it. Also I haven't heard that someone started to work on this. So if you have an idea to fix this issue, PR might be the best approach and will be appreciated.

**Comment by daveyarwood** [comment]
OK -- I'll do some brainstorming and see what I can come up with.

Questions for everyone who has participated in this thread so far ( @inscriber @amartinwest @manishag @virtuald @zhivko @barbaragomes ), so I can better understand the problem?

- How do you want this work, ideally?
- Is interrupting the thread the preferred way to end a thread?
- How do you feel about using shared global state and doing something like this?:

  ```java
  boolean running = true;

  while (running) {
    // receive and handle messages here
  }

  // shut down sockets and destroy context here
  ```

  You could run the `while` loop above in any number of worker threads, and then in the main thread when you want to shut the workers down, you could set `running = false;`.

  Would that solve this problem, or are there pitfalls I'm not aware of?

  If this is a good pattern, then perhaps no code change is needed, and we just need to update our examples to discourage the use of Thread.interrupt().

- Is there any advantage to using Thread.interrupt() instead of doing something like my example above?

- Is the shared global state idea compatible with adding shutdown hooks? In other words, could you add a shutdown hook that sets `running = false;` and somehow wait for all the threads to finish? Would we need some special handling that lets the main thread know when all the worker threads are done, or does this happen automatically?

Any insight on this would be really helpful -- I'm new to this project.

**Comment by daveyarwood** [comment]
I think the point I was missing in my last post is that the work loop can involve blocking calls when receiving, sending or polling. So even if `running` got set to `false`, the worker thread can still potentially be blocked, e.g. waiting for a message.

From what @miniway said, it sounds like the `while (!Thread.currentThread().isInterrupted())` thing we see so often in the examples is not actually the correct way to respond to the context being terminated.

It seems like the correct way to do this is to wrap your blocking calls in a try/catch, capture ZMQExceptions and check if the error code is ETERM.

Quoting @miniway earlier in this issue:

> Other example might be useful for clean exit is
https://github.com/zeromq/jeromq/blob/master/src/test/java/guide/interrupt.java
>
> When the term is called, all the blocking call(usually recv) will raise an Exception, then you can do any cleanup including closing zmq socket.

So I think the work loops in the zguide examples ought to look more like this:

```java
while (true) {
  try {
    // handle messages / do work here
  } catch (ZMQException e) {
    if (e.getErrorCode () == ZMQ.Error.ETERM.getCode ()) {
      break;
    }
  }
}

// if using ZMQ.Context
socket.setLinger(0);
socket.close();
context.term();

// else if using ZContext, which handles closing sockets for you
context.destroy();
```

This still doesn't address the fact that you can't call `.interrupt()` on a blocked worker thread, but unless I'm missing something, I don't see why you would ever need to do that, when it's probably better to do one of these two things instead:

- Terminate the context in the main thread, causing the ETERM exception to get caught and the worker thread to break out of the while loop and shut down.
- Send a shutdown message from the main thread and have the worker handle it by shutting down. (This may be useful if you only want to shut down a particular worker thread, but you still want to keep your ZMQ application running and therefore don't want to terminate the context.)

**Comment by daveyarwood** [comment]
I re-read this thread and I think I fully understand the problem now. 

Clearly it is possible to unblock a worker thread that is blocked on a ZMQ task like waiting to receive a message. You can either send it a shutdown message or terminate its context and let it handle the ETERM exception.

But it becomes problematic when the worker thread gets blocked on something outside of JeroMQ's control. We need some way of interrupting the thread so we don't have to wait for it to finish before the context can terminate. But, we supposedly can't call Thread.interrupt() before terminating the context, so where does that leave us? 

I don't have an answer yet, but I at least thought I'd summarize the problem, in case it's helpful.

**Comment by daveyarwood** [comment]
Open question: what is the undesirable thing that happens when you call Thread.interrupt() before terminating a context?

The JeroMQ wiki says:

> Context.term() must be called before interrupting a thread by Thread.interrupt. Otherwise internal pipe which is used to signal ZMQ commands will be shutdown unexpectedly.

But I have no idea what "internal pipe which is used to signal ZMQ commands will be shutdown unexpectedly" means. It would be useful to know where the problem is in the codebase so we can potentially fix it. Perhaps there is some way we can catch InterruptedException and shut down cleanly.

**Comment by daveyarwood** [comment]
Today I stumbled upon [this](http://www.yegor256.com/2015/10/20/interrupted-exception.html), which is very informative about `Thread.interrupt()`, `Thread.interrupted`, and `InterruptedException` in Java.

I looked through the codebase and found 2 places where we appear to be swallowing an InterruptedException -- I made a PR to fix that: #423.

I don't think that my PR solves this issue, but hopefully it is a step in the right direction.

At this point, I'm thinking the problem is that we aren't checking `Thread.interrupted()` often enough in our code, and there must be at least one place where we should be detecting that the thread is interrupted and taking the opportunity to shut down cleanly, but we aren't.
